Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs

```diff
@@ -0,0 +1,110 @@
+
+namespace Content.Server.Atmos.EntitySystems;
+
+public sealed class GenericGasReactionSystem : EntitySystem
+{
+    [Dependency] private readonly AtmosphereSystem _atmosphere = default!;
+
+    /// <summary>
+    ///     Return a reaction rate (in units reactants per second) for a given reaction. Based on the
+    ///     Arrhenius equation (https://en.wikipedia.org/wiki/Arrhenius_equation).
+    ///
+    ///     This means that most reactions scale exponentially above the MinimumTemperatureRequirement.
+    /// </summary>
+    private float ReactionRate(GasReactionPrototype reaction, GasMixture mix, float dE)
+    {
+        float temp = mix.Temperature;
+
+        // Gas reactions have a MinimumEnergyRequirement which is in spirit activiation energy (Ea),
+        // but no reactions define it. So we have to calculate one to use. One way is to assume that
+        // Ea = 10*R*MinimumTemperatureRequirement such that Ea >> RT.
+        float TScaleFactor = 10;
+        float Ea = TScaleFactor*Atmospherics.R*reaction.MinimumTemperatureRequirement + dE;
+
+        // To compute initial rate coefficient A, assume that at temp = min temp we return 1/10.
+        float RateScaleFactor = 10; // not necessarily the same as TScaleFactor! Don't get confused!
+        float A = MathF.Exp(TScaleFactor) / RateScaleFactor;
+
+        return A*MathF.Exp(-Ea/(Atmospherics.R*temp));
+    }
+
+    /// <summary>
+    ///     Run all of the reactions given on the given gas mixture located in the given container.
+    /// </summary>
+    public ReactionResult ReactAll(IEnumerable<GasReactionPrototype> reactions, GasMixture mix, IGasMixtureHolder? holder)
+    {
+        foreach (var reaction in reactions)
+        {
+            float rate = 1f; // rate of this reaction
+            int reactants = 0;
+
+            // Add concentration-dependent reaction rate
+            // For 1A + 2B -> 3C, the concentration-dependence is [A]^1 * [B]^2
+            float nTotal = mix.TotalMoles;
+            if (nTotal < Atmospherics.GasMinMoles)
+                continue;
+
+            foreach (var (reactant, num) in reaction.Reactants)
+            {
+                rate *= MathF.Pow(mix.GetMoles(reactant)/nTotal, num);
+                reactants++;
+            }
+
+            // No reactants; this is not a generic reaction.
+            if (reactants == 0)
+                continue;
+
+            // Sum catalysts
+            float catalystEnergy = 0;
+            foreach (var (catalyst, dE) in reaction.Catalysts)
+            {
+                catalystEnergy += dE;
+            }
+
+            // Now apply temperature-dependent reaction rate scaling
+            rate *= ReactionRate(reaction, mix, catalystEnergy);
+
+            // Nothing to do
+            if (rate <= 0)
+                continue;
+
+            // Go through and remove all the reactants
+            // If any of the reactants were zero, then the code above would have already set
```

Comment:
This comment is not quite true and for large `rate` we still end up setting negative moles. Working on this locally.